aws/resource_aws_iot_topic_rule_test.go

```diff
@@ -0,0 +1,122 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/service/iot"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAWSIoTTopicRule_basic(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSIoTTopicRuleDestroy_basic,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSIoTTopicRule_basic,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSIoTTopicRuleExists_basic("aws_iot_topic_rule.rule"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAWSIoTTopicRuleDestroy_basic(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).iotconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_iot_topic_rule" {
+			continue
+		}
+
+		out, err := conn.ListTopicRules(&iot.ListTopicRulesInput{})
+
+		if err != nil {
+			return err
+		}
+
+		for _, r := range out.Rules {
+			if *r.RuleName == rs.Primary.ID {
+				return fmt.Errorf("IoT topic rule still exists:\n%s", r)
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func testAccCheckAWSIoTTopicRuleExists_basic(name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		_, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+
+		return nil
+	}
+}
+
+var testAccAWSIoTTopicRule_basic = `
+resource "aws_iot_topic_rule" "rule" {
+  name = "MyRule"
```

Comment:
What do you want the randomization to accomplish? That'll help me determine the "other places" that need it.

Reply:
Generally with testing resources we try to ensure:
* Resources are identifiable by test name
* Parallel runs of a test can be performed by multiple people in the same account
* Dangling resources left by a previously errant test won't prevent others from successfully running their test
* Doing this randomization even if there is an underlying ID since it's good practice and generally code duplication is quite widespread at the moment

Hope this helps!